enterprise/server/hit_tracker_client/hit_tracker_client.go

```diff
@@ -2,57 +2,280 @@ package hit_tracker_client
+		maxPendingHitsPerGroup: *maxPendingHitsPerGroup,
+		hitsByGroup:            map[groupID]*cacheHits{},
+		rpcTimeout:             *remoteHitTrackerTimeout,
+		client:                 hitpb.NewHitTrackerServiceClient(conn),
+	}
+	for i := 0; i < *remoteHitTrackerWorkers; i++ {
+		go factory.start()
+	}
+	env.GetHealthChecker().RegisterShutdownFunction(factory.shutdown)
+	return &factory
+}
+
+type groupID string
+type cacheHits struct {
+	authHeaders map[string][]string
+	hits        []*hitpb.CacheHit
+}
+
 type HitTrackerFactory struct {
-	env environment.Env
+	authenticator interfaces.Authenticator
+
+	ticker <-chan time.Time
+	quit   chan struct{}
+
+	mu                     sync.Mutex
+	maxPendingHitsPerGroup int
+	hitsByGroup            map[groupID]*cacheHits
+
+	rpcTimeout time.Duration
+	client     hitpb.HitTrackerServiceClient
+}
+
+func (h *HitTrackerFactory) NewACHitTracker(ctx context.Context, requestMetadata *repb.RequestMetadata) interfaces.HitTracker {
+	return &NoOpHitTracker{}
+}
+
+func (h *HitTrackerFactory) NewCASHitTracker(ctx context.Context, requestMetadata *repb.RequestMetadata) interfaces.HitTracker {
+	return &HitTrackerClient{ctx: ctx, enqueueFn: h.enqueue, client: h.client, requestMetadata: requestMetadata}
+}
+
+// For Action Cache hit-tracking, explicitly ignore everything. These cache
+// artifacts should always be served from the authoritative cache which will
+// take care of hit-tracking.
+type NoOpHitTracker struct {
+}
+
+func (h *NoOpHitTracker) SetExecutedActionMetadata(md *repb.ExecutedActionMetadata) {
+}
+
+func (h *NoOpHitTracker) TrackMiss(d *repb.Digest) error {
+	return nil
 }
 
-func Register(env *real_environment.RealEnv) {
-	env.SetHitTrackerFactory(HitTrackerFactory{})
+func (h *NoOpHitTracker) TrackDownload(d *repb.Digest) interfaces.TransferTimer {
+	return &NoOpTransferTimer{}
 }
 
-// TODO(iain): hook this up with a to-be-added RPC service in the app.
-type HitTracker struct {
+func (h *NoOpHitTracker) TrackUpload(d *repb.Digest) interfaces.TransferTimer {
+	return &NoOpTransferTimer{}
 }
 
-func (h HitTrackerFactory) NewACHitTracker(ctx context.Context, requestMetadata *repb.RequestMetadata) interfaces.HitTracker {
-	return &HitTracker{}
+type NoOpTransferTimer struct {
 }
 
-func (h HitTrackerFactory) NewCASHitTracker(ctx context.Context, requestMetadata *repb.RequestMetadata) interfaces.HitTracker {
-	return &HitTracker{}
+func (t *NoOpTransferTimer) CloseWithBytesTransferred(bytesTransferredCache, bytesTransferredClient int64, compressor repb.Compressor_Value, serverLabel string) error {
+	return nil
+}
+
+func (t *NoOpTransferTimer) Record(bytesTransferred int64, duration time.Duration, compressor repb.Compressor_Value) error {
+	return nil
 }
 
-func (h *HitTracker) SetExecutedActionMetadata(md *repb.ExecutedActionMetadata) {
+// For CAS hit-tracking, use a hit-tracker that sends information about local
+// cache hits to the RPC service at the configured backend.
```

Comment:
ditto; this comment seems like it'd fit better in NewCASHitTracker

Reply:
Done